python_modules/dagster/dagster/_core/definitions/internal_asset_graph.py

```diff
@@ -163,3 +113,100 @@ def includes_materializable_and_source_assets(self, asset_keys: AbstractSet[Asse
         selected_source_assets = self.source_asset_keys & asset_keys
         selected_regular_assets = asset_keys - self.source_asset_keys
         return len(selected_source_assets) > 0 and len(selected_regular_assets) > 0
+
+    @property
+    @cached_method
+    def asset_dep_graph(self) -> DependencyGraph[AssetKey]:
+        return generate_asset_dep_graph(self._assets_defs, self._source_assets)
+
+    @property
+    def all_asset_keys(self) -> AbstractSet[AssetKey]:
+        return {key for ad in self._assets_defs for key in ad.keys}
+
+    @property
+    def source_asset_keys(self) -> AbstractSet[AssetKey]:
+        return {sa.key for sa in self.source_assets}
+
+    @property
+    def materializable_asset_keys(self) -> AbstractSet[AssetKey]:
+        return {key for ad in self._assets_defs if ad.is_materializable for key in ad.keys}
+
+    def is_materializable(self, asset_key: AssetKey) -> bool:
+        return self.get_assets_def(asset_key).is_materializable
+
+    @property
+    def observable_asset_keys(self) -> AbstractSet[AssetKey]:
+        return {key for ad in self._assets_defs if ad.is_observable for key in ad.keys}
+
+    def is_observable(self, asset_key: AssetKey) -> bool:
+        # Performing an existence check temporarily until we change callsites
+        return self.has_asset(asset_key) and self.get_assets_def(asset_key).is_observable
+
+    @property
+    def external_asset_keys(self) -> AbstractSet[AssetKey]:
+        return {key for ad in self._assets_defs if ad.is_external for key in ad.keys}
+
+    def is_external(self, asset_key: AssetKey) -> bool:
+        # Preserving non-standard behavior of returning True for non-existent keys until callsites
+        # can be updated
+        return asset_key not in self.materializable_asset_keys
+
+    @property
+    def executable_asset_keys(self) -> AbstractSet[AssetKey]:
+        return {key for ad in self._assets_defs if ad.is_executable for key in ad.keys}
+
+    def is_executable(self, asset_key: AssetKey) -> bool:
+        return self.get_assets_def(asset_key).is_executable
+
+    def asset_keys_for_group(self, group_name: str) -> AbstractSet[AssetKey]:
+        return {
+            key
+            for ad in self._assets_defs
+            for key in ad.keys
+            if ad.group_names_by_key[key] == group_name
+        }
+
+    def get_required_multi_asset_keys(self, asset_key: AssetKey) -> AbstractSet[AssetKey]:
```

Comment:
is required a new adjective in our ontology? I figured out based on code (keys that will have corresponding results 100% of the time when the containing op is executed) but it wasn't super natural

Reply:
Nah that's been around for a long time. I usually get confused by it when I encounter it. `get_required_multi_asset_keys` is a method that is already present on existing `AssetGraph`.

Reply:
The word makes more sense to me in the context of an output (since that is always thought of in the context of an op) whereas the word makes less sense whening thinking about an asset in the context of a group or graph.